Add support for Android view capturing in Session Replay

jonathanmos opened the pull request:
### What does this PR do?
Adds support for capturing Android views in Session Replay. Views are supported with the opacity prop or where specifying the opacity through a color with alpha (e.g. FF000000)

Support for border width and color, and for corner radius.

emulator:
<img width="128" alt="Screenshot 2023-11-26 at 21 28 49" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/d4a49e26-2b83-4cc7-a8e5-7fc83a38d9d8">
replay:
<img width="235" alt="Screenshot 2023-11-28 at 12 07 48" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/74f66b65-4756-4ce7-8c32-866a7091ad00">

With border and rounded corners.
emulator:
<img width="342" alt="Screenshot 2023-11-29 at 12 50 55" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/49a3225b-e22d-4243-8653-6507bd9524c1">


replay:
<img width="231" alt="Screenshot 2023-11-29 at 12 50 09" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/00d7a762-4463-49dd-ab38-6b6d97b06bad">


With parent with 10% opacity:
emulator
<img width="346" alt="Screenshot 2023-11-28 at 13 12 52" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/fa3429d7-9051-4e8e-a49f-0bbaae48b9c4">

replay
<img width="244" alt="Screenshot 2023-11-28 at 13 12 42" src="https://github.com/DataDog/dd-sdk-reactnative/assets/48201295/cafb0cb7-bd44-4fa9-b62d-a3ad1f38a62b">

### Motivation
Part of the effort to add Session Replay to React Native.

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

louiszawadzki commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,105 @@
+        val viewAlpha = resolveAlphaValue(view.alpha, backgroundColor)
+
+        if (viewAlpha == 0) {
+            return emptyList()
```
I see 2 corner cases that I'd like to check with this condition:
1. If there is another element inside the view that does not have an alpha of `0`, is it going to be rendered on the screen? if so, is it going to be in the replay as well?
2. If a View has an alpha of `0`, I think it still takes space on the screen? does it on the player as well? (I also think so since I think we use coordinates to place elements)

jonathanmos replied:
With the current changes, 1 is handled and if the parent has an opacity of 0 then the children are not rendered. Regarding the second, we specify wireframes by absolute coordinates so I don't think it will be an issue. Let's discuss a test for this

louiszawadzki commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,105 @@
+        val pixelDensity = mappingContext.systemInformation.screenDensity
+        val backgroundDrawable = view.background as? ReactViewBackgroundDrawable
+
+        val backgroundColor = backgroundDrawable?.color ?: 0
```
Can you attach a screenshot of a view with no background color?
I'm not sure about setting `0` as default value here, I'm afraid it would make the background all black when it should be transparent by default

jonathanmos replied:
It appears that 0 is the code for transparency as defined in Color.java. In recent rn versions they've made this more explicit in the sourcecode

louiszawadzki commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,105 @@
+            return emptyList()
+        }
+
+        val color = colorAndAlphaAsStringHexa(
```
nit: `color` is usually used for font color, maybe we can change it to `backgroundColorAndAlpha`?

louiszawadzki commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,35 @@
+internal fun resolveAlphaValue(viewAlpha: Float, backgroundColor: Int): Int {
+    // must not be toHexString because we'll lose the sign bit
+    val bgColorAsHexCode = backgroundColor.toString(HEX_RADIX)
+    var pct = viewAlpha
```
I don't know what `pct` is, could you make the variable name more explicit or add a comment if it is a concept in color?

jonathanmos replied:
I'll change it to opacityPercentage - a value between 0 and 1

louiszawadzki commented on packages/react-native-session-replay/android/src/test/kotlin/com/datadog/reactnative/sessionreplay/ColorUtilsTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+        // Then
+        assertThat(alphaValue)
+            .isEqualTo(102)
```
Could you explain how `102` is the expected value? You can translate from decimal to hex in comments for more clarity

jonathanmos replied:
I'll add a comment, but basically color code 1717960806 translates to hex 66660066, where 66 is the hex alpha code. [66 hex alpha code is 40%](https://gist.github.com/lopspower/03fb1cc0ac9f32ef38f4) and 40% of 255 is 102 alpha.

louiszawadzki commented:
Can you add screenshots of the nigthly app to see what it looks like?

Also I don't see anything around the `opacity` prop, if this is handled by another prop, could you mention it in a comment in the code?

louiszawadzki commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,113 @@
+        } else {
+            val cornerRadius = backgroundDrawable.fullBorderRadius
+            val borderColor = resolveBorderColor(backgroundDrawable, opacity)
+            val borderWidth = backgroundDrawable.fullBorderWidth
```
On the screenshots you shared I'm under the impression that the border appears bigger in the replay than on emulators. 
Could you check that this works properly?
Maybe we're missing `.convertToDensityNormalized(pixelDensity)`?

jonathanmos replied:
yes, good catch - I've added conversion to dp

louiszawadzki commented on packages/core/android/build.gradle:
```diff
@@ -169,10 +169,10 @@ dependencies {
   implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
   compileOnly "com.squareup.okhttp3:okhttp:3.12.13"
   
-  implementation "com.datadoghq:dd-sdk-android-rum:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-logs:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-trace:2.2.0"
-  implementation "com.datadoghq:dd-sdk-android-webview:2.2.0"
```
Can you make this change part of another commit? that will make it easier to revert if it needs to be reverted.

jonathanmos replied:
done 👍

jonathanmos committed:
Update Android sdk version to 2.3.0

jonathanmos committed:
Implement Android SR for ReactViewGroup

louiszawadzki reviewed (dismissed):
Works for me, I'd like to have an extra approval from a member of the Android SDK team as well.

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,53 @@
+}
+
+@VisibleForTesting
+internal fun alphaCodeToPct(hexAlpha: String): Float {
```
I think we can avoid using the short abbreviation

```suggestion
internal fun alphaCodeToPercentage(hexAlpha: String): Float {
```

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,53 @@
+private const val OPAQUE_ALPHA_VALUE: Int = 255
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveAlpha(opacity: Float, backgroundColor: Int?): Int {
```
I'm a bit confused by this method. It means, that if `backgroundColor` has alpha bits set, we completely ignore `opacity`, it is kind of non-obvious behavior. Why this method takes both parameters and not take only `backgroundColor`, extracts alpha and then compares with `opacity` somewhere else?

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,113 @@
+        viewColor: String,
+        opacity: Float,
+        pixelDensity: Float
+    ):
+            Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```
```suggestion
    ): Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```

jonathanmos commented:
I refactored three files -> ColorUtils, ColorUtilsTest and ReactViewGroupMapper. The changes were quite extensive so could you please review those three again? Basically, I separated opacity and color alpha entirely as concepts and implemented a simpler method to calculate the rgba from the color int. I also removed the recursion for the opacity, because this logic is not used on the native side and the solution will either have to be on the player side, or on the native side perhaps will require an rfc

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,27 @@
+internal fun resolveBackgroundColorAsHex(backgroundColor: Int): String {
+    val colorHexString = Integer.toHexString(backgroundColor)
+
+    // reorder the hex string from argb to rgba
```
I think comment is not needed, method is self-explanatory

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,27 @@
+}
+
+@VisibleForTesting
+internal fun reorderARGBtoRGBA(hexString: String): String {
```
maybe just `argbToRgba`? or `convertArgbToRgba`. I guess these names are more common, at least `convert` is the lingvo used in `android.graphics.Color`.

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,27 @@
+    return "#${reorderARGBtoRGBA(colorHexString)}"
+}
+
+@VisibleForTesting
```
it seems we can make this method private and just test `resolveBackgroundColorAsHex`?

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ColorUtils.kt:
```diff
@@ -0,0 +1,27 @@
+
+private const val HEX_COLOR_INCLUDING_ALPHA_LENGTH: Int = 8
+
+internal fun resolveBackgroundColorAsHex(backgroundColor: Int): String {
```
it is not clear which format will we return. Normally everyone can expect ARGB, but this method returns RGBA. Maybe it makes sense to rename this method to `formatAsRgba` or something similar? Also obviously no need to mention `background` in the name, this method works with any color.

jonathanmos replied:
👍 I'll rename it to formatAsRgba

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,88 @@
+        )
+    }
+
+    @SuppressLint("VisibleForTests")
```
why this annotation is needed? what is the warning?

jonathanmos replied:
The error is "This method should only be accessed from tests or within private scope", returning from view.backgroundColor, because this method is annotated VisibleForTesting

0xnm commented on packages/react-native-session-replay/android/src/main/kotlin/com/datadog/reactnative/sessionreplay/ReactViewGroupMapper.kt:
```diff
@@ -0,0 +1,88 @@
+        backgroundDrawable: ReactViewBackgroundDrawable,
+        opacity: Float,
+        pixelDensity: Float
+    ): Pair<MobileSegment.ShapeStyle?, MobileSegment.ShapeBorder?> {
```
it seems that return type should be `Pair<MobileSegment.ShapeStyle, MobileSegment.ShapeBorder>`, because the only return path in this method return non-null values for the pair.
